polar-core/src/formatting.rs

```diff
@@ -57,36 +54,27 @@ impl Trace {
     }
 }
 
-/// Traverse a [`Source`](../types/struct.Source.html) line by line until `offset` is reached,
-/// and return the source line containing the `offset` character as well as `num_lines` lines
-/// above and below it.
+/// Traverse a [`Source`](../types/struct.Source.html) line by line until `offset` is reached and
+/// return the source line containing the `offset` character as well as `num_lines` lines above and
+/// below it.
 // @TODO: Can we have the caret under the whole range of the expression instead of just the beginning.
 pub fn source_lines(source: &Source, offset: usize, num_lines: usize) -> String {
-    // Sliding window of lines: current line + indicator + additional context above + below.
-    let max_lines = num_lines * 2 + 2;
-    let push_line = |lines: &mut Vec<String>, line: String| {
-        if lines.len() == max_lines {
-            lines.remove(0);
-        }
-        lines.push(line);
-    };
-    let mut index = 0;
-    let mut lines = Vec::new();
-    let mut target = None;
+    let (target_line, target_column) = loc_to_pos(&source.src, offset);
+    // Skip everything up to the first line of requested context (`target_line - num_lines`), but
+    // don't overflow if `num_lines > target_line`.
+    let skipped_lines = target_line.saturating_sub(num_lines);
```

Comment:
TIL about `saturating_sub`. neat.

Reply:
I tripped up reading this at first because I had in my head that `num_lines` was the _total number_ of lines to be returned, rather than the number of _lines of context either side_ of the match to be returned. maybe `s/num_lines/context_lines/` ?

Reply:
Same! I was overflowing (underflowing? lol) and needed to figure out a way to spell "subtract down to zero but no further" in Rust

Reply:
> I tripped up reading this at first because I had in my head that `num_lines` was the _total number_ of lines to be returned, rather than the number of _lines of context either side_ of the match to be returned. maybe `s/num_lines/context_lines/` ?

That's great feedback; thanks — I'll push a change.

Reply:
5a4a8643